src/App.ts

```diff
@@ -252,11 +253,50 @@ export default class App {
           signingSecret,
           endpoints,
           processBeforeResponse,
+          clientId,
+          clientSecret,
+          stateStore,
+          stateSecret,
+          installationStore,
+          authVersion,
+          metadata,
+          scopes,
           logger: this.logger,
         });
       }
     }
 
+    let usingBuiltinOauth = undefined;
+    if (
+      clientId !== undefined
+      && clientSecret !== undefined
+      && (stateSecret !== undefined || stateStore !== undefined)
+      && this.receiver instanceof ExpressReceiver
+    ) {
+      usingBuiltinOauth = true;
+    }
+
+    if (token !== undefined) {
+      if (authorize !== undefined || usingBuiltinOauth !== undefined) {
+        throw new AppInitializationError(
+          `token as well as authorize options or installer options were provided. ${tokenUsage}`,
+        );
+      }
+      this.authorize = singleTeamAuthorization(this.client, { botId, botUserId, botToken: token });
+    } else if (authorize === undefined && usingBuiltinOauth === undefined) {
+      throw new AppInitializationError(
+        `No token, no authorize options, and no installer options provided. ${tokenUsage}`,
+      );
+    } else if (authorize !== undefined && usingBuiltinOauth !== undefined) {
+      throw new AppInitializationError(
+        `Both authorize options and installer options provided. ${tokenUsage}`,
+      );
+    } else if (authorize === undefined && usingBuiltinOauth !== undefined) {
+      this.authorize = (this.receiver as ExpressReceiver).oauthAuthorize as Authorize;
+    } else if (authorize !== undefined && usingBuiltinOauth === undefined) {
+      this.authorize = authorize;
+    }
```

Comment:
we could use an exhaustiveness check here to enforce that there are no fall-through cases: https://www.typescriptlang.org/docs/handbook/advanced-types.html#exhaustiveness-checking